File: JetpacReloaded/Typescript version/ts/bonus.ts
import { Color } from "../ts_lib/color.js";
import { Rectangle } from "../ts_lib/rectangle.js";
import { SpriteBatch, Texture2D } from "../ts_lib/spritebatch.js";
import { Vector2 } from "../ts_lib/vector2.js";
import { BaseObject } from "./baseobject.js";

export class Bonus extends BaseObject {
    private bonusLanded: boolean = false;
    private m_prevFrame: number = 0;

    constructor(image: Texture2D) {
        super();
        this.m_frame = this.rand.Next(0, 4);
        this.m_image = image;
        this.m_width = image.width / 5;
        this.m_height = image.height + 2;
        this.m_screenLocation = new Vector2(this.rand.Next(0, 750), -30);
    }

    public Update(): void {
        if (!this.bonusLanded) {
            this.m_screenLocation.Y++;
        }
        this.m_rect = new Rectangle(this.m_frame * this.m_width, 0, this.m_width, this.m_height);
    }
    public Draw(spriteBatch: SpriteBatch): void {
        let bonusLocation: Vector2 = new Vector2(this.m_screenLocation.X, this.m_screenLocation.Y);
        spriteBatch.Draw(this.m_image, bonusLocation, this.m_rect, Color.White, 0, Vector2.Zero, 1.0, 0, 0);
    }
    public get BonRect(): Rectangle {
        return new Rectangle(this.m_screenLocation.X, this.m_screenLocation.Y, this.m_width, this.m_height);
    }
    public set BonusLanded(value: boolean) {
        this.bonusLanded = value;
    }
    public Reset(): void {
        this.m_prevFrame = this.m_frame;
        this.m_frame = this.rand.Next(0, 4);
        if (this.m_frame == this.m_prevFrame) {
            this.m_frame = (this.m_frame + 1) % 4;
        }
        this.m_screenLocation.X = this.rand.Next(0, 750);
        this.m_screenLocation.Y = -30;
        this.bonusLanded = false;
    }
}

File: JetpacReloaded/C# version/JetpacReloaded/Bonus.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

namespace JetPacReloaded
{
    internal class Bonus : BaseObject
    {
        private bool _bonusLanded = false;
        private int _mPrevFrame = 0;

        public Bonus(Texture2D image)
        {
            m_frame = rand.Next(0, 4);
            m_image = image;
            m_width = image.Width / 5;
            m_height = image.Height + 2;
            m_screenLocation = new Vector2(rand.Next(0, 750), -30);
        }

        public void Update()
        {
            if (!_bonusLanded)
            {
                m_screenLocation.Y++;
            }
            m_rect = new Rectangle(m_frame * m_width, 0, m_width, m_height);
        }

        public new void Draw(SpriteBatch spriteBatch)
        {
            Vector2 bonusLocation = new Vector2(m_screenLocation.X, m_screenLocation.Y);
            spriteBatch.Draw(m_image, bonusLocation, m_rect, Color.White, 0, Vector2.Zero, 1.0f, 0, 0);
        }

        public Rectangle BonRect => new((int)m_screenLocation.X, (int)m_screenLocation.Y, m_width, m_height);

        public bool BonusLanded { set => _bonusLanded = value; }

        public void Reset()
        {
            _mPrevFrame = m_frame;
            m_frame = rand.Next(0, 4);
            if (m_frame == _mPrevFrame)
            {
                m_frame = (m_frame + 1) % 4;
            }
            m_screenLocation.X = rand.Next(0, 750);
            m_screenLocation.Y = -30;
            _bonusLanded = false;
        }
    }
}
